chisel/src/executor.rs

```diff
@@ -376,82 +395,129 @@ impl Type {
     ///
     /// Optionally, an owned [Type]
     fn from_expression(expr: &pt::Expression) -> Option<Self> {
-        Some(match expr {
-            pt::Expression::Type(_, ty) => match ty {
-                pt::Type::Address | pt::Type::AddressPayable | pt::Type::Payable => {
-                    Self::Builtin(ParamType::Address)
-                }
-                pt::Type::Bool => Self::Builtin(ParamType::Bool),
-                pt::Type::String => Self::Builtin(ParamType::String),
-                pt::Type::Int(size) => Self::Builtin(ParamType::Int(*size as usize)),
-                pt::Type::Uint(size) => Self::Builtin(ParamType::Uint(*size as usize)),
-                pt::Type::Bytes(size) => Self::Builtin(ParamType::FixedBytes(*size as usize)),
-                pt::Type::DynamicBytes => Self::Builtin(ParamType::Bytes),
-                pt::Type::Mapping(_, _, right) => Self::from_expression(right)?,
-                _ => return None,
-            },
-            pt::Expression::Variable(ident) => Self::Custom(vec![ident.name.clone()]),
+        match expr {
+            pt::Expression::Type(_, ty) => Self::from_type(ty),
+
+            pt::Expression::Variable(ident) => Some(Self::Custom(vec![ident.name.clone()])),
+            pt::Expression::This(_) => Some(Self::Custom(vec!["this".to_string()])),
+
+            // array
             pt::Expression::ArraySubscript(_, expr, num) => {
-                let num = num.as_ref().and_then(|num| {
-                    if let pt::Expression::NumberLiteral(_, num, exp) = num.as_ref() {
-                        let num = if num.is_empty() { 0usize } else { num.parse().ok()? };
-                        let exp = if exp.is_empty() { 0u32 } else { exp.parse().ok()? };
-                        Some(num * 10usize.pow(exp))
+                Self::from_expression(expr).and_then(|ty| {
+                    let ty = Box::new(ty);
+                    let num = num.as_deref().and_then(parse_number_literal);
+                    if let Some(num) = num {
+                        // overflow check
+                        const MAX_U64_AS_U256: U256 = U256([0, 0, 0, usize::MAX as u64]);
+                        if num > MAX_U64_AS_U256 {
```

Comment:
last time we changed U256 constants in ethers we got burned, so I'd appreciate a simple sanity check for this constant.